feat(tangle-dapp): Add Rewards Claiming Functionality

AtelyPham opened the pull request:
## Summary of changes

_Provide a detailed description of proposed changes._

- Display locked balance horizontally to free up space for additional information.
- Added a reward claiming button to the account summary card.
- Relocated Tangle asset ID types to the `types` directory.
- Implemented claim rewards logic for Substrate and EVM.
- Added a hook to connect action event handlers with UI states.

### Proposed area of change

_Put an `x` in the boxes that apply._

- [x] `apps/tangle-dapp`
- [ ] `apps/tangle-cloud`
- [x] `libs/tangle-shared-ui`
- [ ] `libs/webb-ui-components`

### Associated issue(s)

_Specify any issues that can be closed from these changes (e.g. `Closes #233`)._

- Closes https://github.com/tangle-network/dapp/issues/2726
- Related https://github.com/tangle-network/dapp/issues/2727

### Screen Recording

_If possible provide screenshots and/or a screen recording of proposed change._

**_Substrate_**

https://github.com/user-attachments/assets/4ae8d96a-14db-48d3-8b83-ba314773d804

**_EVM_**

https://github.com/user-attachments/assets/9386e635-c815-4afc-a81b-0343bd61ef40

AtelyPham commented:
The reward-claiming test in the video failed because there are no active services to increase the rewards, resulting in no rewards to claim. 

In the demo, I enabled the claiming function in the code to demonstrate interaction with the chains (Substrate & EVM). In reality, we have a check to render the claiming button.

monaiuu commented:
@AtelyPham I thought we don't need active services to claim the reward, users should be eligible for TNT rewards with just making a deposit in restaking. 

also can we update the UI as the following so it is clear to the user the amount of unclaimed rewards?

![image](https://github.com/user-attachments/assets/069bfe9a-6378-4a2a-aec4-dbe83abf3976)

AtelyPham commented:
We still need space for these actionsbuttons. How would we place the content? 

![CleanShot 2025-01-23 at 16 27 20@2x](https://github.com/user-attachments/assets/e9249e5b-43e5-41da-9222-160f16f9d163)

monaiuu reviewed (commented):
we can move the send action button next to the TNT amount (shown in image above), as for viewing the rewards currently that is for nomination. so I think user can just go to nomination page to claim rewards. we don't have to collide it with claiming restaking rewards.

monaiuu commented:
as for vest we are currently conditionally rendering it anyway, and having it here already:

<img width="1512" alt="Screenshot 2025-01-23 at 4 31 54 PM" src="https://github.com/user-attachments/assets/c8f1d5c7-e355-44be-986e-be365dccbbb2" />

monaiuu commented:
@1xstj can you set up the rewards up for testnet restake vaults so we can get this PR through?

drewstone commented on apps/tangle-dapp/src/components/account/ClaimRewardAction.tsx:
```diff
@@ -0,0 +1,124 @@
+
+  // TODO: Add other assets in the future,
+  // for now, we only support native asset
+  const claimableAssets = useMemo<TangleAssetId[]>(() => {
```
Need to actually iterate over all assets the user has non-zero deposits in in vaults and then batch the calls / chunk them so that we can make efficient tx processing here.

drewstone reviewed (changes requested):
Seems our reward processing is not where we need it to be right now. We can't leave todos like this for later. These features need to ship 100% complete. No more breaking features like this down it's bad practice for this type of thing and is slowing us down.

drewstone commented:
Status update? There are now EVM assets and you and yuri should have ability to create them to test out rewards, points, and more.

AtelyPham commented:
After my meeting with Mona yesterday, there were some UI updates, and new APIs were added to retrieve the details of ERC20 assets. It will be ready for review by the end of today.

AtelyPham commented:
### Changes since the last review:

- Moved actions next to the account balance to create space for reward and point displays.
- Added error handling for unsupported pallets or API calls.
- Removed the `View` button from the operator table.
- Changed the `Restake` button to `Delegate` on the operator table.
- Added a blueprint tab on the restaking page.
- Removed the join operator banner from the restaking page.
- Moved the join operator button to the operator tab on the restaking page.
- Update copies of tooltips.
- Switched to the delegate tab and auto-selected the operator when the user clicks the `delegate` button on the operator table.
- Updated the Substrate asset ID to `bigint` from `number`.
- Added support for batching multiple asset rewards simultaneously, if needed.

yurixander commented on apps/tangle-dapp/src/components/account/RewardsAndPoints.tsx:
```diff
@@ -0,0 +1,117 @@
+      const representativeAsset = Object.entries(result).find(
+        ([assetId, balance]) => {
+          const hasBalance = balance > ZERO_BIG_INT;
+          const hasMetadata = assetsMetadata.get(assetId as RestakeAssetId);
```
Avoid casting and instead use runtime assertions to catch future logic bugs which may otherwise go silent and become harder to debug:
```suggestion
          const hasMetadata = assetsMetadata.get(assertRestakeAssetId(assetId));
```

yurixander commented on apps/tangle-dapp/src/components/account/RewardsAndPoints.tsx:
```diff
@@ -0,0 +1,117 @@
+      }
+
+      const [assetId, balance] = representativeAsset;
+      const metadata = assetsMetadata.get(assetId as RestakeAssetId);
```
Avoid casting and instead use runtime assertions to catch future logic bugs which may otherwise go silent and become harder to debug

yurixander commented on apps/tangle-dapp/src/components/account/RewardsAndPoints.tsx:
```diff
@@ -0,0 +1,117 @@
+        return null;
+      }
+
+      return `${formatUnits(balance, metadata.decimals)} ${metadata.symbol}`;
```
Prefer using `formatDisplayAmount(balance, metadata.decimals, AmountFormatStyle.SHORT)` for improved consistency & handles some special cases where the balance has a lot of decimal places.

yurixander commented on apps/tangle-dapp/src/components/account/RewardsAndPoints.tsx:
```diff
@@ -0,0 +1,117 @@
+      (substrateNativeBalance ?? ZERO_BIG_INT) +
+      (evmNativeBalance ?? ZERO_BIG_INT);
+
+    const formattedBalance = formatUnits(
+      totalBalance,
+      activeChain.nativeCurrency.decimals,
+    );
```
Prefer using `formatDisplayAmount(balance, metadata.decimals, AmountFormatStyle.SHORT)` for improved consistency & handles some special cases where the balance has a lot of decimal places.

yurixander commented on apps/tangle-dapp/src/components/account/RewardsAndPoints.tsx:
```diff
@@ -0,0 +1,117 @@
+          Object.keys(result).length > 0 &&
+          displayRewardAsset ? (
+            <ClaimRewardAction
+              claimableAssetIds={Object.keys(result) as RestakeAssetId[]}
```
```suggestion
              claimableAssetIds={Object.keys(result).map(assertRestakeAssetId)}
```

yurixander commented on libs/tangle-shared-ui/src/types/error.ts:
```diff
@@ -21,3 +21,7 @@ export class TangleError extends Error {
   }
 }
+
+export function isTangleError(error: unknown): error is TangleError {
+  return error instanceof TangleError;
+}
```
It may be simpler to simply use `if (error instanceof TangleError)` directly without this function

yurixander commented on libs/tangle-shared-ui/src/types/index.ts:
```diff
@@ -119,3 +120,7 @@ export type BridgeTokenWithBalance = BridgeToken & {
+
+export type RestakeAssetId = `${bigint}` | EvmAddress;
+
+export type TangleAssetId = { Custom: u128 } | { Erc20: Address };
```
So that we have a bit stronger types when creating asset IDs for transactions:
```suggestion
export type TangleAssetId = { Custom: u128 } | { Erc20: EvmAddress };
```

yurixander commented on libs/webb-ui-components/src/utils/isTemplateBigInt.ts:
```diff
@@ -0,0 +1,17 @@
+    return false;
+  }
+
+  if (!/^[+-]?\d+$/.test(value)) {
+    return false;
+  }
```
Should be able to skip this check if we're creating the BigInt instance down below? 🤔

AtelyPham replied:
This check validates the hex string, ensuring it contains only digit characters within the bigint template.
